torch/_dynamo/_trace_wrapped_higher_order_op.py

```diff
@@ -0,0 +1,109 @@
+import functools
+
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import autograd_not_implemented
+
+from torch._ops import HigherOrderOperator
+from torch._subclasses import FakeTensorMode
+
+from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode, track_tensor_tree
+from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+
+# Trace wrapped is a higher order op meant for both invoking a bound function,
+# and for registering it as a call_function in the backward graph.
+# This allows us to re-enter dynamo during compiled autograd to trace (or graph break)
+# the functions as needed. This, in turn, means we can support functions in backward with complex python
+# state mutation. If we were to not do this, the functions would get inlined into their composing aten ops,
+# and we would lose the python state mutation.
```

Comment:
Because we don't have meta functions for this yet, you need to more clearly specify what the invariants of functions you can validly call are. I believe the function must take a single input and return an output with exactly the same tensor metadata as the original (but it's allowed to be a new tensor).

There's also something subtle going on with whether or not it's a new tensor or not. If there was possibility for mutation in backwards, it would be necessary to know for meta-purposes if you got the new tensor or not. However, there is not (because backwards only ever mutates into accumulate grads), so you get off here? IDK this all needs to be spelled out.

Reply:
Hmm, are custom `autograd.Function`'s allowed to mutate grad_outputs in the backward? (even if this only allowed in eager then this is still a problem, because compiled autograd will capture the entire `backward.)`

Reply:
No, they are not. We protect this with a context.